Web Inspector: At some narrow-ish widths objects and disclosure triangles in console output disappear

the-chenergy opened the pull request:
#### 7a3f0bd5f4ea707c63069210b01f8674ce2a5509
<pre>
Web Inspector: At some narrow-ish widths objects and disclosure triangles in console output disappear
<a href="https://rdar.apple.com/121861970">rdar://121861970</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=274630">https://bugs.webkit.org/show_bug.cgi?id=274630</a>

Reviewed by Devin Rousso and Patrick Angle.

The object display, including the disclosure triangles, are instances
of ObjectTreeBaseTreeElement&apos;s subclasses, which have the
`.object-tree` class name as DOM elements. This commit improves the
styles of these elements by making them always `display: inline` since
it worked best with the text truncating feature that the elements have.

While making the `.object-tree` elements `display: inline`, the tree
element was rendered incorrectly when displayed as an indexed item,
with a preceding array index or a label like &quot;key&quot; or &quot;value&quot;.
The fix is to use a `display: flex` for the entry row instead, so
the `.index-name` (the label) and the `.index-value` (containing the
`.object-tree`) are always side by side instead of the `.index-value`
being wrapped and shown on a second line.

This commit also make it so that lines that are cut off in the object
tree elements always consistently show ellipses at the end. In addition,
we take out the existing unnecessary right padding on the tree elements
to make the cut off point vertically aligned.

* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
(.console-top-level-message .object-tree): Deleted.
(.console-top-level-message .object-tree .object-tree): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
(.object-tree):
  - Always use `display: inline` for the object tree element. It works
    well with the text-truncating styles the element has.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
(.tree-outline.object &gt; .item):
(.tree-outline.object):
  - Remove the unnecessary right padding.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
(.object-tree-array-index &gt; .titles &gt; .title):
  - Use `display: flex` so that the name and the value elements are
    always side by side instead of the value wrapped to a second line
    when it gets wide.

(.object-tree-array-index .index-name): Deleted.
(.object-tree-array-index &gt; .titles &gt; .title &gt; .index-name):
  - Use `min-width` instead of `width` to go with the `display: flex`
    in the parent.
  - Refine the selector to make it more precise; there might be nested
    object trees inside, and we don&apos;t actually want this style to
    control those `.index-name` elements beyond the one controlled by
    the current object tree.

(.object-tree-array-index &gt; .titles &gt; .title &gt; .index-value):
  - Make ellipses show up for these lines.

(.tree-outline .item.object-tree-array-index):
  - Move the `left` style into the
    `.object-tree-array-index &gt; .titles &gt; .title` selector instead so
    that we expand the element&apos;s viewable area rather than moving the
    element. (Moving the element resulted in the right edge being
    shifted to the left and made the cut off point not align vertically
    with the parent&apos;s right margin.)

(.tree-outline .item.object-tree-array-index:not(.object-tree-map-entry)):
  - The map entry element does not need the left expansion like the
    array entry. Undo the `margin-left` style.

(.object-tree-property + ol .object-tree-array-index): Deleted.
  - This is no longer required.

* Source/WebInspectorUI/UserInterface/Views/LocalRemoteObjectContentView.css:
(.content-view.local-remote-object):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.css:
(.popover .debugger-popover-content &gt; .body):
  - The object trees within these containers were allowed to overflow
    horizontally and were scrollable to the right. The `display: inline`
    change would make them not scrollable anymore. So, we make these
    containers that expect the object tree child to be scrollable
    `display: flex`, and we must commit to doing the same to all
    such containers of scrollable object trees in the future.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeMapEntryTreeElement.css:
(.object-tree-array-index.object-tree-map-entry &gt; .titles &gt; .title &gt; .index-name):
  - Use `min-width` instead of `width` to go with the `display: flex`
    in the parent.

Canonical link: <a href="https://commits.webkit.org/279894@main">https://commits.webkit.org/279894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a616289aa76f0264ae927ab67980dd9a6a92c34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/54526 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/33943 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/7094 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57804 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/5258 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/41484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/5261 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/57804 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/5258 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/56622 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/41484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/7094 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57804 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/41484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/7094 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/3399 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/41484 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/7094 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/59395 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/5261 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/59395 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30906 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/7094 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/59395 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12050 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31888 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30696 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
```diff
@@ -25,7 +25,7 @@
 .object-tree {
     position: relative;
-    display: inline-block;
+    display: inline;
```
This will apply to every object tree throughout Web Inspector, including the ones used in the Sources tab for the Scope sidebar as well as in the popovers when hovering over a value paused in the debugger. We should confirm those still behave and size as expected with this adjustment.

If not, it seems like it would still be valid to set `display: inline;` for `.console-top-level-message .object-tree`.

the-chenergy replied:
Good idea. I went ahead and tested the object tree view in the following places:

(All screenshots are "after versus before")

<img width="1873" alt="Screenshot 2024-05-31 at 11 57 11" src="https://github.com/WebKit/WebKit/assets/44916353/d78a0a21-dc23-4aa3-a95f-306e06620a24">
<img width="1758" alt="Screenshot 2024-05-31 at 11 15 30" src="https://github.com/WebKit/WebKit/assets/44916353/61ad6192-7925-4eea-be51-77a09d238590">
<img width="1700" alt="Screenshot 2024-05-31 at 11 16 43" src="https://github.com/WebKit/WebKit/assets/44916353/ee5fb1f7-e24b-4b59-bed8-121d3bedfa41">
<img width="1926" alt="Screenshot 2024-05-31 at 11 46 40" src="https://github.com/WebKit/WebKit/assets/44916353/372bd76c-1d2b-4407-a556-acab9b52abe3">

As seen from the screenshots, this patch also appropriately addresses the same issue with object trees displayed elsewhere in the inspector.

However, there are two notable side effects brought by this patch:

1. Some cut-off lines now show ellipses at the end. I think this is fine because ellipses should generally be expected on lines that don't fit and are unscrollable. Though as it stands there are still lines without ellipses which is rather inconsistent, so this side effect is barely an improvement.
2. In the last two screenshots, the original behavior was that the object trees were scrollable to the right, and this patch made them not scrollable anymore. **I would consider this a regression,** so I think what we can do is let the parents of the object tree decide whether it should be scrollable to the right, by optionally being `display: flex` to allow scrolling.

dcrousso replied:
in the case that something isn't horizontally scrollable, i agree that showing "…" is probably better than just letting the content keep going, but in all of those screenshots it looks like the content is cut off _way_ before the end of the container, which is definitely not ideal

unless that's fixed by your most recent changes?

the-chenergy replied:
Good catch! I added a patch for that so that the cut-off point for any nested text is always at the same verticle boundary, basically by taking out all the `padding-right`s those tree elements used to have.

dcrousso replied:
btw i have a more general question

when/why did all this break?  i know this UI was fine for **_years_**, so it's extremely odd for it to suddenly have issues.  makes me wonder if Web Inspector was doing something hacky or if WebKit has changed such that suddenly there's web compatibility issues :/

the-chenergy replied:
Right, I agree that it might be worthwhile to pinpoint what happened to cause that strange drop of portions of the text as it might reveal something more problematic. I will find some time to do that later and just apply the current fix for now.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
```diff
@@ -84,7 +84,7 @@
 .tree-outline.object {
     margin: 0;
-    padding: 0 6px 2px;
+    padding: 0 0 2px 6px;
```
i wanna say that even though the Console is always LTR, there are places where object trees are shown RTL (e.g. hovering over a token when JavaScript is paused the Sources Tab)

if that's the case, do these still work when in RTL?

the-chenergy replied:
What specific test case do you have in mind, like what's an instance of a RTL object tree? The popover with the object tree looks fine to me in the Sources tab. Sure, it's without the 6px right padding when scrolled all the way to the right, but the difference is then when the tree is nested, there won't be a large stack of right padding applied on the right. Given there's no left padding, a right padding shouldn't be necessary.

dcrousso replied:
i dont have a specific situation in mind, more of trying to follow the general principle of having the UI look exactly the same when in RTL as it does when in LTR

> it's without the 6px right padding when scrolled all the way to the right

this makes me think that maybe we should have the CSS be something like this instead?

```css
padding-top: 0;
padding-inline: 6px 0;
padding-bottom: 2px;
```

put another way, whenever i see `padding-left`/`padding-right` i often think "should this be `padding-inline-start`/`padding-inline-end` instead?"

the-chenergy replied:
Ah, I see what you mean. Let's use `padding-inline` instead and avoid having to worry about the `left` and `right` ness.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeOutline.css:
```diff
@@ -47,7 +47,7 @@
 
-    padding: 0 var(--tree-outline-item-padding);
+    padding: 0 0 0 var(--tree-outline-item-padding);
 
```
this seems like a **_very_** scary change since it could affect _all_ trees everywhere in Web Inspector

are we sure this is OK for all the various uses?  could we maybe just override this for object trees instead?

the-chenergy replied:
Whoops, I did not notice that I'm changing `.tree-outline` instead of `.object-tree` at this point. Very scary indeed 😨

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
```diff
@@ -34,7 +33,12 @@
-.object-tree-array-index .index-name {
+.object-tree-array-index > .titles > .title {
+    display: flex;
+    margin-left: -12px;
```
i think this should be `margin-inline-start`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
```diff
@@ -45,20 +49,21 @@
     display: inline-block;
-    width: 30px;
+    min-width: 30px;
     text-align: right;
```
(i realize that this isn't entirely related to the PR)

i think this should be `text-align: end`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
```diff
@@ -45,20 +49,21 @@
 
+.object-tree-array-index > .titles > .title > .index-value {
+    overflow: hidden;
+    margin-left: 8px;
```
i think this should be `margin-inline-start`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeMapEntryTreeElement.css:
```diff
@@ -23,8 +23,12 @@
  */
 
+.object-tree-array-index.object-tree-map-entry > .titles > .title {
+    margin-left: 0;
```
i think this should be `margin-inline-start`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
```diff
@@ -45,20 +49,21 @@
 
     display: inline-block;
-    width: 30px;
+    min-width: 30px;
```
so if we have a really long array, would this potentially mean that the space taken up by the index is large enough that the value for that index is shifted to the side (relative to all the other values)?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeMapEntryTreeElement.css:
```diff
@@ -23,8 +23,12 @@
+
 .object-tree-array-index.object-tree-map-entry > .titles > .title > .index-name {
-    width: 40px;
+    min-width: 40px;
```
so if we have a really long array, would this potentially mean that the space taken up by the index is large enough that the value for that index is shifted to the side (relative to all the other values)?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
```diff
@@ -84,12 +84,18 @@
 }
 
+.tree-outline.object > .item {
+    padding-right: 0;
```
i think this should be `padding-inline-end`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeMapEntryTreeElement.css:
```diff
@@ -23,8 +23,12 @@
 .object-tree-array-index.object-tree-map-entry > .titles > .title > .index-name {
-    width: 40px;
+    min-width: 40px;
     text-align: right;
```
(i realize that this isn't entirely related to the PR)

i think this should be `text-align: end`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
```diff
@@ -25,7 +25,6 @@
 
 .tree-outline .item.object-tree-array-index {
     position: relative;
-    left: -12px;
```
i think this might actually cause the index to be cut off when it's large enough (similar to my concern below about using `min-width`)

the-chenergy replied:
I reintroduced `left: -12px` (as `inset-inline-start: -12px` to be RTL-compatible). The reason I used `margin-left` instead was to fix the extra spacing that it caused on the right. I offset that with `margin-inline-end: -12px`.

dcrousso reviewed (commented):
nice work!  overall this looks really good, but i think there's some spots that need RTL support :)

id suggest looking through all the files you've modified and finding anything with `right`/`left`/etc. and replacing it with the corresponding property/value that's independent of direction (though there are a few spots where i think it's fine to leave as-is, such as for things containing a square `background-image`)

if you weren't aware, you can force the direction to be RTL by
1. do a local build
2. open Web Inspector
3. press ⇧⌥⌘D to enable Web Inspector debug settings
4. go to the Settings Tab
5. go to the Debug pane
6. use the Layout Direction dropdown

dcrousso reviewed (approved):
nice work!  thanks for fixing some of the other RTL issues along the way :)

the-chenergy committed:
Web Inspector: At some narrow-ish widths objects and disclosure triangles in console output disappear
rdar://121861970
https://bugs.webkit.org/show_bug.cgi?id=274630

Reviewed by Devin Rousso and Patrick Angle.

The object display, including the disclosure triangles, are instances
of ObjectTreeBaseTreeElement's subclasses, which have the
`.object-tree` class name as DOM elements. This commit improves the
styles of these elements by making them always `display: inline` since
it worked best with the text truncating feature that the elements have.

While making the `.object-tree` elements `display: inline`, the tree
element was rendered incorrectly when displayed as an indexed item,
with a preceding array index or a label like "key" or "value".
The fix is to use a `display: flex` for the entry row instead, so
the `.index-name` (the label) and the `.index-value` (containing the
`.object-tree`) are always side by side instead of the `.index-value`
being wrapped and shown on a second line.

This commit also make it so that lines that are cut off in the object
tree elements always consistently show ellipses at the end. In addition,
we take out the existing unnecessary right padding on the tree elements
to make the cut off point vertically aligned.

* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
(.console-top-level-message .object-tree): Deleted.
(.console-top-level-message .object-tree .object-tree): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
(.object-tree):
  - Always use `display: inline` for the object tree element. It works
    well with the text-truncating styles the element has.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeView.css:
(.tree-outline.object > .item):
(.tree-outline.object):
  - Remove the unnecessary right padding.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeArrayIndexTreeElement.css:
(.object-tree-array-index > .titles > .title):
  - Use `display: flex` so that the name and the value elements are
    always side by side instead of the value wrapped to a second line
    when it gets wide.

(.object-tree-array-index .index-name): Deleted.
(.object-tree-array-index > .titles > .title > .index-name):
  - Use `min-width` instead of `width` to go with the `display: flex`
    in the parent.
  - Refine the selector to make it more precise; there might be nested
    object trees inside, and we don't actually want this style to
    control those `.index-name` elements beyond the one controlled by
    the current object tree.

(.object-tree-array-index > .titles > .title > .index-value):
  - Make ellipses show up for these lines.

(.tree-outline .item.object-tree-array-index):
  - Move the `left` style into the
    `.object-tree-array-index > .titles > .title` selector instead so
    that we expand the element's viewable area rather than moving the
    element. (Moving the element resulted in the right edge being
    shifted to the left and made the cut off point not align vertically
    with the parent's right margin.)

(.tree-outline .item.object-tree-array-index:not(.object-tree-map-entry)):
  - The map entry element does not need the left expansion like the
    array entry. Undo the `margin-left` style.

(.object-tree-property + ol .object-tree-array-index): Deleted.
  - This is no longer required.

* Source/WebInspectorUI/UserInterface/Views/LocalRemoteObjectContentView.css:
(.content-view.local-remote-object):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.css:
(.popover .debugger-popover-content > .body):
  - The object trees within these containers were allowed to overflow
    horizontally and were scrollable to the right. The `display: inline`
    change would make them not scrollable anymore. So, we make these
    containers that expect the object tree child to be scrollable
    `display: flex`, and we must commit to doing the same to all
    such containers of scrollable object trees in the future.

* Source/WebInspectorUI/UserInterface/Views/ObjectTreeMapEntryTreeElement.css:
(.object-tree-array-index.object-tree-map-entry > .titles > .title > .index-name):
  - Use `min-width` instead of `width` to go with the `display: flex`
    in the parent.

Canonical link: https://commits.webkit.org/279894@main
